pkg/controller/node.go

```diff
@@ -98,23 +100,41 @@ func (r *ReconcileNode) Recover(ctx context.Context) error {
 	ctx, w := workflow.New(ctx)
 	defer func() { w.Finish(err) }()
 
-	var nodes *azdiskv1beta2.AzDriverNodeList
-	if nodes, err = r.azClient.DiskV1beta2().AzDriverNodes(r.config.ObjectNamespace).List(ctx, metav1.ListOptions{}); err != nil {
-		if errors.IsNotFound(err) {
-			return nil
-		}
+	var azNodes *azdiskv1beta2.AzDriverNodeList
+	if azNodes, err = r.azClient.DiskV1beta2().AzDriverNodes(r.config.ObjectNamespace).List(ctx, metav1.ListOptions{}); err != nil {
 		return err
 	}
 
-	for _, node := range nodes.Items {
-		updated := node.DeepCopy()
-		updated.Annotations = azureutils.AddToMap(updated.Annotations, consts.RecoverAnnotation, "azDriverNode")
-		if _, err = r.azClient.DiskV1beta2().AzDriverNodes(r.config.ObjectNamespace).Update(ctx, updated, metav1.UpdateOptions{}); err != nil {
-			return err
+	for _, azNode := range azNodes.Items {
+		// if the corresponding node has been deleted, delete the azdrivernode object
+		_, err := r.kubeClient.CoreV1().Nodes().Get(ctx, azNode.Spec.NodeName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				n := r.azClient.DiskV1beta2().AzDriverNodes(r.config.ObjectNamespace)
+				if err = n.Delete(ctx, azNode.Name, metav1.DeleteOptions{}); err != nil {
+					w.Logger().Errorf(err, "failed to delete azDriverNode (%s)", azNode.Name)
+				}
+			} else {
+				w.Logger().Errorf(err, "failed to find node (%s)", azNode.Spec.NodeName)
+			}
+		} else {
+			updateFunc := func(obj client.Object) error {
+				azNode := obj.(*azdiskv1beta2.AzDriverNode)
+				azNode.Annotations = azureutils.AddToMap(azNode.Annotations, consts.RecoverAnnotation, "azDriverNode")
+				return nil
+			}
+			if _, err = azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, &azNode, updateFunc, consts.NormalUpdateMaxNetRetry, azureutils.UpdateCRI); err != nil {
+				w.Logger().Errorf(err, "failed to recover azDriverNode (%s) with annotation", azNode.Name)
+				continue
+			}
+			r.addNodeToAvailableAttachmentsMap(ctx, azNode.Name, azNode.GetLabels())
 		}
-		r.addNodeToAvailableAttachmentsMap(ctx, node.Name, node.GetLabels())
 	}
-	return nil
+
+	if err != nil {
```

Comment:
You'll need to count the errors separately since `err` contains the result of the last iteration. An error won't be returned if the last iteration is successful but earlier iterations failed.

Reply:
Actually, I just noticed that `err` was recreated (using `:=`) in the loop scope, so `err` will *always* `== nil` at this point.